samples/net/prometheus/src/main.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/net/http/server.h>
+#include <zephyr/net/http/service.h>
+#include <zephyr/net/net_ip.h>
+#include <zephyr/net/socket.h>
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/counter.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+
+struct {
+
+	struct prometheus_collector *collector;
+
+	struct prometheus_counter *counter;
+
+} prom_context;
+
+#if defined(CONFIG_NET_SAMPLE_HTTP_SERVICE)
+static uint16_t test_http_service_port = CONFIG_NET_SAMPLE_HTTP_SERVER_SERVICE_PORT;
+HTTP_SERVICE_DEFINE(test_http_service, CONFIG_NET_CONFIG_MY_IPV4_ADDR, &test_http_service_port, 1,
+		    10, NULL);
+
+static uint8_t recv_buffer[2048];
+
+static int dyn_handler(struct http_client_ctx *client, enum http_data_status status,
+		       uint8_t *buffer, size_t len, void *user_data)
+{
+#define MAX_TEMP_PRINT_LEN 32
+	int rc;
+	char hex_len[MAX_TEMP_PRINT_LEN];
+	static char print_str[MAX_TEMP_PRINT_LEN];
+	enum http_method method = client->method;
+	static size_t processed;
+	static bool replied;
+
+	__ASSERT_NO_MSG(buffer != NULL);
+
+	if (status == HTTP_SERVER_DATA_ABORTED) {
+		LOG_DBG("Transaction aborted after %zd bytes.", processed);
+		processed = 0;
+		replied = false;
+		return 0;
+	}
+
+	processed += len;
+
+	snprintf(print_str, sizeof(print_str), "%s received (%zd bytes)", http_method_str(method),
+		 len);
+	LOG_HEXDUMP_DBG(buffer, len, print_str);
+
+	if (status == HTTP_SERVER_DATA_FINAL) {
+		LOG_DBG("All data received (%zd bytes).", processed);
+		processed = 0;
+	}
+
+	if (strlen(buffer) == 0) {
+		replied = false;
+	}
+
+	/* incrase counter per request */
+	prometheus_counter_inc(prom_context.counter);
+
+	(void)memset(recv_buffer, 0, sizeof(recv_buffer));
+
+	prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
```

Comment:
This function can fail but it returns `void`, and is called in a function that returns `int`.

How about taking advantage of your existing error checking effort to connect the error reporting further? :)

```suggestion
	err = prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
	if (err < 0) {
		return err;
	}
```